core/git.go

```diff
@@ -0,0 +1,101 @@
+package core
+
+import (
+	"context"
+
+	"github.com/dagger/dagger/core/pipeline"
+	"github.com/dagger/dagger/core/socket"
+	"github.com/dagger/dagger/engine"
+	"github.com/dagger/dagger/engine/buildkit"
+	"github.com/dagger/dagger/engine/sources/gitdns"
+	"github.com/moby/buildkit/client/llb"
+	specs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+type GitRef struct {
+	URL string `ssh:"url"`
+	Ref string `ssh:"ref"`
+
+	KeepGitDir bool `ssh:"keepGitDir"`
+
+	SSHKnownHosts string    `json:"sshKnownHosts"`
+	SSHAuthSocket socket.ID `json:"sshAuthSocket"`
```

Comment:
nit: we can generally store the actual object (`*Socket`) rather than the ID, which is slightly more convenient to use most of the time. Not that important either way though, just an fyi

Reply:
Interesting - it looks like the most of the rest of core also just stores the ID? Is this something we could refactor as well?

I'll leave as is for now (but will happily do a follow-up as a general clean-up) :tada:

Reply:
I also prefer just storing the object directly. We started with storing IDs, but for no good reason, we just didn't stop and think about it at the time. It's nice having the object readily available without having to decode it, and the ID stores all of the same data anyway at the moment, just with extra steps in between.

We could maybe reconsider this if/when v2 IDs are a thing and IDs are no longer literally the encoded object, but eh, I think even in that case we might just want to keep these as objects, which might be a literal pointer reference to the same object stored in the ID cache.

Reply:
Just to note: I've just copied the property from where it was before, this isn't a new one: https://github.com/dagger/dagger/blob/6a72a0951500599e8c406a71ac92fabf1019e35b/core/schema/git.go#L118-L121

Not sure if that changes anything, happy to try and fix it if we're moving it (but would prefer not to block on it if possible).

I tried looking at changing the type, but I'm a bit confused - we're using the ID directly: https://github.com/dagger/dagger/blob/39c00b6fd7c473572634a78e52e9272e965ca093/core/git.go#L69-L71

If we store the object directly, how do we pass it to `MountSSHSocket`?

Reply:
Ha, you'd have to call `ID()` and pass it in there, which is probably why this hasn't been tweaked - we actually _want_ it in ID form because we have to shim that value through Buildkit to our socket provider where it's "re-hydrated" from ID.

Fine with leaving this as-is, none of this is blocking, mostly just taking the opportunity to talk about this since I've wanted to bring it up numerous times before but it's just never been very important.